aws_lambda_powertools/utilities/streaming/s3_object.py

```diff
@@ -0,0 +1,235 @@
+        in_place: bool, optional
+            Transforms the stream in place, instead of returning a new stream object. Defaults to false.
+
+        Returns
+        -------
+        T[bound=IO[bytes]], optional
+            If in_place is False, returns an IO[bytes] object representing the transformed stream
+        """
+        if self.tell() != 0:
+            raise ValueError(f"Cannot add transformations to a read object. Already read {self.tell()} bytes")
+
+        # Make transformations always be a sequence to make mypy happy
+        if not isinstance(transformations, Sequence):
+            transformations = [transformations]
+
+        if in_place:
+            self._data_transformations.extend(transformations)
+
+            # Invalidate any existing transformed stream.
+            # It will be created again next time it's accessed.
+            self._transformed_stream = None
+            return None
+        else:
+            # Tell mypy that raw_stream actually implements T (bound to IO[bytes])
+            stream = cast(T, self.raw_stream)
+            for transformation in transformations:
+                stream = transformation.transform(stream)
+            return stream
+
+    # From this point on, we're just implementing all the standard methods on the IO[bytes] type.
+    # There's no magic here, just delegating all the calls to our transformed_stream.
+    def seek(self, offset: int, whence: int = io.SEEK_SET) -> int:
+        return self.transformed_stream.seek(offset, whence)
+
+    def seekable(self) -> bool:
+        return self.transformed_stream.seekable()
+
+    def readable(self) -> bool:
+        return self.transformed_stream.readable()
+
+    def writable(self) -> bool:
+        return self.transformed_stream.writable()
+
+    def tell(self) -> int:
+        return self.transformed_stream.tell()
+
+    @property
+    def closed(self) -> bool:
+        return self.transformed_stream.closed
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self.close()
+
+    def close(self):
+        self.raw_stream.close()
+
+        # Also close transformed stream if there are any transformations
+        if self.raw_stream != self._transformed_stream and self._transformed_stream is not None:
+            self._transformed_stream.close()
+
+    def read(self, size: int = -1) -> bytes:
+        return self.transformed_stream.read(size)
+
+    def readline(self, size: Optional[int] = -1) -> bytes:
+        return self.transformed_stream.readline()
+
+    def readlines(self, hint: int = -1) -> List[bytes]:
+        return self.transformed_stream.readlines(hint)
+
+    def __next__(self):
+        return self.transformed_stream.__next__()
+
+    def __iter__(self):
+        return self.transformed_stream.__iter__()
+
+    def fileno(self) -> int:
+        raise NotImplementedError()
+
+    def flush(self) -> None:
+        raise NotImplementedError()
```

Comment:
Add brief docstring to quickly warn them that this is not available before they crash at runtime (for those with autocomplete but Mypy)